tiledb/sm/filesystem/ls_scanner.h

```diff
@@ -0,0 +1,232 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This defines the LsScanner class and related types used for VFS.
+ */
+
+#ifndef TILEDB_LS_SCANNER_H
+#define TILEDB_LS_SCANNER_H
+
+#include "tiledb/sm/filesystem/uri.h"
+
+#include <cstdint>
+#include <functional>
+#include <stdexcept>
+
+/** Inclusion predicate for objects collected by ls */
+template <class F>
+concept FilePredicate = true;
+
+/**
+ * DirectoryPredicate is currently unused, but is kept here for adding directory
+ * pruning support in the future.
+ */
+template <class F>
+concept DirectoryPredicate = true;
+
+namespace tiledb::sm {
+using FileFilter = std::function<bool(const std::string_view&, uint64_t)>;
+
+using DirectoryFilter = std::function<bool(const std::string_view&)>;
+/** Static DirectoryFilter used as default argument. */
+[[maybe_unused]] static bool accept_all_dirs(const std::string_view&) {
+  return true;
+}
+
+/** Type defintion for objects returned from ls_recursive. */
+using LsObjects = std::vector<std::pair<std::string, uint64_t>>;
+
+/**
+ * LsScanIterator iterates over the results of ls requests wrapped by classes
+ * deriving from LsScanner. See S3Scanner as an example.
+ *
+ * @tparam T The LsScanner type that created this iterator.
+ * @tparam U The data type stored by this iterator.
+ */
+template <class T, class U>
+class LsScanIterator {
+ public:
+  using value_type = U;
+  using difference_type = ptrdiff_t;
+  using pointer = const U*;
+  using reference = const U&;
+  using iterator_category = std::input_iterator_tag;
+
+  /** Default constructor. */
+  LsScanIterator() noexcept = default;
+
+  /**
+   * Constructor.
+   *
+   * @param scanner The scanner that created this iterator.
+   */
+  explicit LsScanIterator(T* scanner) noexcept
+      : scanner_(scanner)
+      , ptr_(scanner->begin_) {
+  }
+
+  /**
+   * Constructor.
+   *
+   * @param scanner The scanner that created this iterator.
+   * @param ptr Pointer to set as the current object.
+   */
+  LsScanIterator(T* scanner, pointer ptr) noexcept
+      : scanner_(scanner)
+      , ptr_(ptr) {
+  }
+
+  /** Copy constructor. */
+  LsScanIterator(const LsScanIterator& rhs) noexcept {
+    ptr_ = rhs.ptr_;
+    scanner_ = rhs.scanner_;
+  }
+
+  /** Copy assignment operator. */
+  LsScanIterator& operator=(LsScanIterator rhs) noexcept {
+    if (&rhs != this) {
+      std::swap(ptr_, rhs.ptr_);
+      std::swap(scanner_, rhs.scanner_);
+    }
+    return *this;
+  }
+
+  /**
+   * Dereference operator.
+   *
+   * @return The current object being visited.
+   */
+  constexpr reference operator*() const noexcept {
+    return *ptr_;
```

Comment:
This will segfault if executed immediately after the default constructor. Because the value of `ptr_` is part of the representation of the end-iterator, there needs to be a check here. It should throw when the iterator is not dereferenceable.